### PR TITLE
schedstatistics: Log sleep tick statistics

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR modifies the schedstatistics module to also log the time sleeping when the idle thread is removed. The time spent sleeping is stored in the entry for `KERNEL_PID_UNDEF`. 

### Testing procedure

Without idle thread and with schedstatistics, `ps` shows the relative time spend on each thread when the CPU is running:

```
> ps
2020-12-15 16:34:14,950 #       pid | name                 | state    Q | pri | stack  ( used) ( free) | base addr  | current     | runtime  | switches
2020-12-15 16:34:14,955 #         - | isr_stack            | -        - |   - |    512 (  196) (  316) | 0x20000000 | 0x200001c0
2020-12-15 16:34:14,961 #         1 | main                 | running  Q |   7 |   1536 (  812) (  724) | 0x200007d8 | 0x20000be4  | 54.586% |       145
2020-12-15 16:34:14,966 #         2 | usbus                | bl anyfl _ |   1 |   1024 (  432) (  592) | 0x20000288 | 0x20000594  | 47.630% |       193
2020-12-15 16:34:14,968 #           | SUM                  |            |     |   3072 ( 1440) ( 1632)
```

 With this PR, the time sleeping is also logged, the numbers now show the percentage of time spend on each thread including the sleep time in this calculation:

```
> ps
2020-12-15 16:38:07,756 #       pid | name                 | state    Q | pri | stack  ( used) ( free) | base addr  | current     | runtime  | switches
2020-12-15 16:38:07,761 #         - | isr_stack            | -        - |   - |    512 (  196) (  316) | 0x20000000 | 0x200001c0
2020-12-15 16:38:07,768 #         1 | main                 | running  Q |   7 |   1536 (  812) (  724) | 0x200007d8 | 0x20000be4  |  0.033% |        42
2020-12-15 16:38:07,774 #         2 | usbus                | bl anyfl _ |   1 |   1024 (  432) (  592) | 0x20000288 | 0x20000594  |  0.045% |        90
2020-12-15 16:38:07,777 #           | SUM                  |            |     |   3072 ( 1440) ( 1632)
```

I've tested this with `tests/usbus_cdc_acm_stdio` on the samr21-xpro.

### Issues/PRs references

None

### CODE REVIEW

**Comment by benpicco** [comment]
my only concern is that those numbers will approach zero pretty fast for most applications

**Comment by bergzand** [comment]
> my only concern is that those numbers will approach zero pretty fast for most applications

Yeah, that's also the case when the idle thread is enabled. The output behavior with this PR should match the numbers when the idle thread is enabled, except that the idle thread is not displayed.

**Comment by bergzand** [comment]
I fixed this because I tried to debug the USB peripheral on the stm32f4. It was showing 99.8% runtime. I was under the impression that it was somehow looping fast between IRQ's and events. In the end it was the schedstatistics module not taking the idle time into account.

**Comment by kaspar030** [comment]
> The output behavior with this PR should match the numbers when the idle thread is enabled, except that the idle thread is not displayed.

can we show this somehow?

File: sys/schedstatistics/schedstatistics.c [hunk 1]
```diff
@@ -26,20 +26,24 @@
 #include "thread.h"
 #include "xtimer.h"
 
+/**
+ * When core_idle_thread is active, the KERNEL_PID_UNDEF is used to track the
```
Comment (kaspar030) [inline sys/schedstatistics/schedstatistics.c:30]: nice one!

**Comment by bergzand** [comment]
I've adapted `tests/bench_mutex_pingpong` to call `ps()` 1 second after the test finished. `test_utils_interactive_sync` is disabled.

With Idle thread:
```
2020-12-15 19:27:50,323 # main(): This is RIOT! (Version: 2021.01-devel-1579-ge9f8c-pr/schedstatistics/remove_idle_runtime)
2020-12-15 19:27:50,324 # main starting
2020-12-15 19:27:51,328 # { "result" : 18182, "ticks" : 2639 }
2020-12-15 19:27:52,339 #       pid | name                 | state    Q | pri | stack  ( used) ( free) | base addr  | current     | runtime  | switches
2020-12-15 19:27:52,348 #         - | isr_stack            | -        - |   - |    512 (  208) (  304) | 0x20000000 | 0x200001c0
2020-12-15 19:27:52,359 #         1 | idle                 | pending  Q |  15 |    256 (  148) (  108) | 0x20000878 | 0x200008e4  | 60.611% |         1
2020-12-15 19:27:52,370 #         2 | main                 | running  Q |   7 |   1536 (  568) (  968) | 0x20000978 | 0x20000e74  | 20.652% |     18185
2020-12-15 19:27:52,381 #         3 | second_thread        | bl mutex _ |   6 |   1536 (  148) ( 1388) | 0x20000278 | 0x200007e4  | 18.735% |     18183
2020-12-15 19:27:52,388 #           | SUM                  |            |     |   3840 ( 1072) ( 2768)
```

Without idle thread:
```
2020-12-15 19:28:16,425 # main(): This is RIOT! (Version: 2021.01-devel-1579-ge9f8c-pr/schedstatistics/remove_idle_runtime)
2020-12-15 19:28:16,426 # main starting
2020-12-15 19:28:17,430 # { "result" : 17544, "ticks" : 2735 }
2020-12-15 19:28:18,442 #       pid | name                 | state    Q | pri | stack  ( used) ( free) | base addr  | current     | runtime  | switches
2020-12-15 19:28:18,451 #         - | isr_stack            | -        - |   - |    512 (  240) (  272) | 0x20000000 | 0x200001c0
2020-12-15 19:28:18,462 #         1 | main                 | running  Q |   7 |   1536 (  568) (  968) | 0x20000878 | 0x20000d9c  | 17.223% |     17547
2020-12-15 19:28:18,473 #         2 | second_thread        | bl mutex _ |   6 |   1536 (  148) ( 1388) | 0x20000278 | 0x200007e4  | 15.684% |     17545
2020-12-15 19:28:18,479 #           | SUM                  |            |     |   3584 (  956) ( 2628)
```

The difference between `main` and `second_thread` is likely from the time the main thread spent on setting up the benchmark. I don't know where the slight difference between the two measurements comes from.

File: sys/schedstatistics/schedstatistics.c (hunk 1 above)
Comment (fjmolinas) [reply sys/schedstatistics/schedstatistics.c:30]: Don't you mean not active?

**Comment by fjmolinas** [comment]
> The difference between main and second_thread is likely from the time the main thread spent on setting up the benchmark. I don't know where the slight difference between the two measurements comes from.

Isn't that simply explained by the difference in context switches?

**Review by fjmolinas — APPROVED** [review]
ACK

**Comment by bergzand** [comment]
> Isn't that simply explained by the difference in context switches?

Very likely, but I haven't verified if that's the case.

**Comment by fjmolinas** [comment]
GO

**Pull request merged**